src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/TaskCompletionSourceWithCancellation.cs

```diff
@@ -0,0 +1,32 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Net.Http
+{
+    internal sealed class TaskCompletionSourceWithCancellation<T> : TaskCompletionSource<T>
+    {
+        private CancellationToken _cancellationToken;
+
+        public TaskCompletionSourceWithCancellation() : base(TaskCreationOptions.RunContinuationsAsynchronously)
+        {
+        }
+
+        private void OnCancellation()
+        {
+            TrySetCanceled(_cancellationToken);
+        }
+
+        public async Task<T> WaitWithCancellationAsync(CancellationToken cancellationToken)
+        {
+            _cancellationToken = cancellationToken;
+            using (cancellationToken.Register(s => ((TaskCompletionSourceWithCancellation<T>)s).OnCancellation(), this))
+            {
+                return await Task.ConfigureAwait(false);
+            }
+        }
+    }
+}
```

Comment:
Have you considered adding a newline at EOF?

Reply:
Added, thanks.